### ISSUE TITLE
Frozen the statistics of BN?

### ISSUE THREAD

**Issue by nihaomiao** [isht7/pytorch-deeplab-resnet#39]
Hello, do you implement the "frozen BN layers"? I find that running means still keep changing during training.

**Comment by isht8** [comment]
running mean and variance would be constant because we use eval mode during training. (see [here](https://github.com/isht7/pytorch-deeplab-resnet/blob/master/train.py#L200))
The parameters of the batchnorm layer have no gradient because of [this setting](https://github.com/isht7/pytorch-deeplab-resnet/blob/master/deeplab_resnet.py#L75) everywhere for batchnorm in model definition

**Comment by nihaomiao** [comment]
Hello, setting model.eval() during training can fix running means and vars. But now I am a bit confused since this setting can also affect other layers. Do other layers can be trained and updated in setting "model.eval()"?

**Comment by isht7** [comment]
model.eval() only affects Batchnorm and dropout layer. Read more about this in pytorch docs. model.eval() is safe here because, there is no dropout layer in our model.

**Comment by nihaomiao** [comment]
Thank you! Now I know that model.eval() doesn't turn off the history tracking..... Thanks a lot!

**Comment by isht7** [comment]
to stop pytorch from storing compute graph, use torch.no_grad()

**Comment by episodeyang** [comment]
@isht8 The running mean and variance does change even under eval mode. The batch norm in pytorch does not freeze these stats unless you tell it explicitly so.
